x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationNodeService.java

```diff
@@ -402,12 +402,14 @@ private void handleLoadFailure(TrainedModelDeploymentTask task, Exception ex) {
                 task.stoppedReason().orElse("_unknown_")
             ));
         }
-        // TODO: Do we want to remove from the modelIdToTask map? This would cause it to be reloaded by state updates on INITIALIZING
-        modelIdToTask.remove(task.getModelId());
+        // TODO: Do we want to stop the task? This would cause it to be reloaded by state updates on INITIALIZING
+        // We should stop the local task so that future task actions won't get routed to the older one.
```

Comment:
Actions should ever get routed to this node as its routing state was never set to `STARTED`. It went from `STARTING` then an attempt to load the model and is now `FAILED`. 

But, I can see the point in removing the model from task manager so that if this node is accidentally added to the tasks request, we don't match against it. What this callback path will do is:

 - Set the stored state to failed (or maybe fail to even do that)
 - THEN remove the task from the modelIdToTask map

This fixes a race condition that might occur previous actually. Line 406 could have been called, then a new cluster state update came in and another attempt to load the model could have been fired because of line 319.

👍